### PR TITLE
#6876 Remote SQLite driver

### LINKED ISSUE #6876: How to connect DBeaver to sqlite3 db through ssh tunnel
#### System information: 
- Client Windows 10 / Server Raspbian 10 rpi3
- DBeaver version 6.2.1 release 23/09/2019

#### Connection specification:
- Database SQLITE3 through SSH

#### Describe your question:
Hi there!
I'm trying to connect on sqlite3 database through ssh on remote server.
My ssh tunnel is working, but what is the syntax and where paste it on the db.sqlite3 file in DBeaver.
DBeaver connection informations :
image

SSH connection informations:
Annotation 2019-09-30 173456
image

### PR DESCRIPTION
**Pull request by ShadelessFox**
image

Separate tickets must be issued for:
1. Performing synchronization during disconnect
2. Refreshing associated editors when loading remote changes from server

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/ui/actions/datasource/DataSourceSynchronizeHandler.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+        if (context == null) {
+            return null;
+        }
+
+        final var dataSource = context.getDataSource();
+        final var dataSourceContainer = dataSource.getContainer();
+        final var dataSourceProvider = (DBPDataSourceProviderSynchronizable) dataSourceContainer.getDriver().getDataSourceProvider();
+
+        new AbstractJob("Synchronize data source [" + dataSource.getName() + "]") {
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/ui/actions/datasource/DataSourceSynchronizeHandler.java:53]:
```suggestion
        new AbstractJob("Synchronize data source [" + dataSource.getName() + "]") {  //$NON-NLS-1$
```

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/DBPDataSourceProviderSynchronizable.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+package org.jkiss.dbeaver.model;
+
+import org.jkiss.code.NotNull;
+import org.jkiss.dbeaver.DBException;
+import org.jkiss.dbeaver.model.runtime.DBRProgressMonitor;
+
+public interface DBPDataSourceProviderSynchronizable extends DBPDataSourceProvider {
+
+    /**
+     * Synchronized the local data source with the remote data source.
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/DBPDataSourceProviderSynchronizable.java]:
```suggestion
     * Synchronizes the local data source with the remote data source.
```

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/DBPDataSourceProviderSynchronizable.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+     * @param container data source container that needs synchronization
+     * @throws DBException on any database error
+     */
+    void syncLocalDataSource(
+        @NotNull DBRProgressMonitor monitor,
+        @NotNull DBPDataSourceContainer container
+    ) throws DBException;
+
+    /**
+     * Synchronized the remote data source with the local data source.
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/DBPDataSourceProviderSynchronizable.java]:
```suggestion
     * Synchronizes the remote data source with the local data source.
```

File: plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceDescriptor.java [hunk 4]
```diff
@@ -950,7 +950,22 @@ public void openDataSource(DBRProgressMonitor monitor, boolean initialize) throw
...
-        this.dataSource = getDriver().getDataSourceProvider().openDataSource(monitor, this);
+
+        final DBPDataSourceProvider provider = getDriver().getDataSourceProvider();
+        if (provider instanceof DBPDataSourceProviderSynchronizable) {
+            try {
+                monitor.beginTask("Synchronize local data source", 1);
+                ((DBPDataSourceProviderSynchronizable) provider).syncLocalDataSource(monitor, this);
+                monitor.worked(1);
+            } catch (DBException e) {
+                DBWorkbench.getPlatformUI().showError("Data source synchronization error", "Error synchronizing local data source", e);
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceDescriptor.java]: Please, add localization or  //$NON-NLS-1$

File: plugins/org.jkiss.dbeaver.net.ssh.sshj/src/org/jkiss/dbeaver/model/net/ssh/SSHImplementationSshj.java [hunk 5]
```diff
@@ -171,6 +176,61 @@ public void invalidateTunnel(DBRProgressMonitor monitor) throws DBException, IOE
...
+                public InputStream getInputStream() {
+                    return src;
+                }
+            }, dst);
+        }
+    }
+
+    @NotNull
+    private SFTPClient openSftpClient() throws DBException, IOException {
+        final SSHClient client = this.sshClient;
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.net.ssh.sshj/src/org/jkiss/dbeaver/model/net/ssh/SSHImplementationSshj.java]: I'm curious why we can't use this field directly.

File: plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceDescriptor.java [hunk 6]
```diff
@@ -1039,6 +1054,23 @@ private boolean disconnect(final DBRProgressMonitor monitor, boolean reflect) {
...
+            final DBPDataSourceProvider provider = driver.getDataSourceProvider();
+            if (provider instanceof DBPDataSourceProviderSynchronizable) {
+                final DBPDataSourceProviderSynchronizable remoteProvider = (DBPDataSourceProviderSynchronizable) provider;
+                if (!remoteProvider.isLocalDataSourceSynchronized(monitor, this)) {
+                    try {
+                        monitor.beginTask("Synchronize remote data source", 1);
+                        remoteProvider.syncRemoteDataSource(monitor, this);
+                        monitor.worked(1);
+                    } catch (DBException e) {
+                        DBWorkbench.getPlatformUI().showError("Data source synchronization error", "Error synchronizing remote data source", e);
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceDescriptor.java]: And here, and in other places..

File: plugins/org.jkiss.dbeaver.net.ssh.sshj/src/org/jkiss/dbeaver/model/net/ssh/SSHImplementationSshj.java (hunk 5 above)
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.net.ssh.sshj/src/org/jkiss/dbeaver/model/net/ssh/SSHImplementationSshj.java]: This field looks like redundant.

**Comment by flambchap** [comment]
Great stuff. I'm excited to try this out

File: plugins/org.jkiss.dbeaver.net.ssh.sshj/src/org/jkiss/dbeaver/model/net/ssh/SSHImplementationSshj.java [hunk 7]
```diff
@@ -171,6 +176,61 @@ public void invalidateTunnel(DBRProgressMonitor monitor) throws DBException, IOE
...
+        }
+    }
+
+    @Override
+    public void putFile(@NotNull InputStream src, @NotNull String dst, @NotNull DBRProgressMonitor monitor) throws DBException, IOException {
+        try (SFTPClient client = openSftpClient()) {
+            client.put(new InMemorySourceFile() {
+                @Override
+                public String getName() {
+                    return "memory";
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.net.ssh.sshj/src/org/jkiss/dbeaver/model/net/ssh/SSHImplementationSshj.java]:
```suggestion
                    return "memory";  //$NON-NLS-1$
```

**Comment by uslss** [comment]
related issues:
https://github.com/dbeaver/pro/issues/1173
https://github.com/dbeaver/pro/issues/1174
https://github.com/dbeaver/pro/issues/1175

**Pull request merged**